components/sup/src/manager/service/config.rs

```diff
@@ -227,6 +234,18 @@ impl ServiceConfig {
         self.needs_write = false;
         Ok(should_restart)
     }
+
+    pub fn reload_package(&mut self, package: &PackageInstall) -> Result<()> {
+        self.config_root = match self.config_root_from_config {
+            Some(ref root) => root.clone(),
+            None => package.installed_path.clone(),
+        };
+        self.pkg = Pkg::new(package,
+                            self.pkg.svc_user.clone(),
+                            self.pkg.svc_group.clone())?;
+        self.cfg = Cfg::new(package, &self.config_root)?;
+        Ok(())
```

Comment:
Before exiting this function I think you want to throw away the entire hook table and load it again. Hooks aren't just additive, they could also be removed between package updates. Doing this here might also simplify some logic below - you probably don't even need to ask for the hooks to be compiled since rebuilding the entire table will clear the incarnation counter and they'll be recompiled on the next call to `run_hook()`